src/core/net/tcp6.cpp

```diff
@@ -302,27 +456,222 @@ Error Tcp::Listener::Deinitialize(void)
+        {
+            otLinkedBuffer *next = curr->mNext;
+            curr->mNext          = nullptr;
+            aEndpoint.mSendDoneCallback(&aEndpoint, curr);
+            curr = next;
+        }
+    }
+    if (aSignals.conn_established && aEndpoint.mEstablishedCallback != nullptr)
+    {
+        aEndpoint.mEstablishedCallback(&aEndpoint);
+    }
+    if ((aSignals.recvbuf_notempty || aSignals.rcvd_fin) && aEndpoint.mReceiveAvailableCallback != nullptr)
+    {
+        aEndpoint.mReceiveAvailableCallback(&aEndpoint, cbuf_used_space(&aEndpoint.mTcb.recvbuf),
+                                            aEndpoint.mTcb.reass_fin_index != -1,
+                                            cbuf_free_space(&aEndpoint.mTcb.recvbuf));
+    }
+    if (aEndpoint.mTcb.t_state == TCP6S_TIME_WAIT && aEndpoint.mDisconnectedCallback != nullptr)
+    {
+        aEndpoint.mDisconnectedCallback(&aEndpoint, OT_TCP_DISCONNECTED_REASON_TIME_WAIT);
+    }
+}
+
+Error Tcp::BsdErrorToOtError(int aBsdError)
+{
+    switch (aBsdError)
+    {
+    case 0:
+        return kErrorNone;
+    default:
+        return kErrorFailed;
+    }
+}
+
+bool Tcp::CanBind(const SockAddr &aSockName)
+{
+    uint16_t port = HostSwap16(aSockName.mPort);
+    for (Endpoint *endpoint = mEndpoints.GetHead(); endpoint != nullptr; endpoint = endpoint->GetNext())
+    {
+        if (endpoint->mTcb.lport == port)
+        {
+            if (aSockName.GetAddress().IsUnspecified())
+            {
+                return false;
+            }
+            if (reinterpret_cast<Address *>(&endpoint->mTcb.laddr)->IsUnspecified())
+            {
+                return false;
+            }
+            if (memcmp(&endpoint->mTcb.laddr, &aSockName.mAddress, sizeof(endpoint->mTcb.laddr)) == 0)
+            {
+                return false;
+            }
+        }
+    }
+
+    for (Listener *listener = mListeners.GetHead(); listener != nullptr; listener = listener->GetNext())
+    {
+        if (listener->mTcbListen.lport == port)
+        {
+            if (aSockName.GetAddress().IsUnspecified())
+            {
+                return false;
+            }
+            if (reinterpret_cast<Address *>(&listener->mTcbListen.laddr)->IsUnspecified())
+            {
+                return false;
+            }
+            if (memcmp(&listener->mTcbListen.laddr, &aSockName.mAddress, sizeof(listener->mTcbListen.laddr)) == 0)
+            {
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
+bool Tcp::AutoBind(const SockAddr &aPeer, SockAddr &aToBind, bool aBindAddress, bool aBindPort)
+{
+    if (aBindAddress)
+    {
+        MessageInfo peerInfo;
+        peerInfo.Clear();
+        peerInfo.SetPeerAddr(aPeer.GetAddress());
+        const Netif::UnicastAddress *netifAddress =
+            InstanceLocator::GetInstance().Get<Ip6>().SelectSourceAddress(peerInfo);
+        if (netifAddress == nullptr)
+        {
+            return false;
```

Comment:
single return statement? (can use `ExitNow()`)

Reply:
OK. Fixed in the latest update.